internal/templates/containercapabilities/template.go

```diff
@@ -0,0 +1,218 @@
+	"golang.stackrox.io/kube-linter/internal/templates"
+	"golang.stackrox.io/kube-linter/internal/templates/containercapabilities/internal/params"
+	"golang.stackrox.io/kube-linter/internal/templates/util"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	reservedCapabilitiesAll             = "all"
+	matchLiteralReservedCapabilitiesAll = "(?i)" + reservedCapabilitiesAll
+)
+
+var (
+	literalReservedCapabilitiesAllMatcher func(string) bool
+)
+
+func checkCapabilityDropList(
+	containerName string,
+	paramCapMatchers map[string]func(string) bool,
+	scCaps *v1.Capabilities,
+	forbidAll bool,
+) *diagnostic.Diagnostic {
+	if forbidAll {
+		// Specified to drop all. Instead of trying to find a match to this paramCap, we should
+		// make sure the special ReservedCapabilitiesAll is found in the DROP list as well
+		for _, scCap := range scCaps.Drop {
+			if literalReservedCapabilitiesAllMatcher(string(scCap)) {
+				// Found "all" in drop list as well
+				return nil
+			}
+		}
+		return &diagnostic.Diagnostic{
+			Message: fmt.Sprintf(
+				"container %q has DROP capabilities: %q, but in fact all capabilities are required to be dropeed",
+				containerName,
+				scCaps.Drop),
+		}
+	}
+
+	// Every forbidden capability specified by param should be found in the DROP list
+	for paramCap, paramCapMatcher := range paramCapMatchers {
+		var found bool
+		for _, scCap := range scCaps.Drop {
+			// User can specify to drop "all" under containers as well
+			if paramCapMatcher(string(scCap)) || literalReservedCapabilitiesAllMatcher(string(scCap)) {
+				// This forbidden cap exists in the DROP list. Check for the next forbidden cap
+				found = true
+				break
+			}
+		}
+		if !found {
+			return &diagnostic.Diagnostic{
+				Message: fmt.Sprintf("container %q has DROP capabilities: %q, but does not drop "+
+					"capability %q which is required",
+					containerName,
+					scCaps.Drop,
+					paramCap),
+			}
+		}
+	}
+	return nil
+}
+
+func checkCapabilityAddList(
+	containerName string,
+	paramCapMatchers map[string]func(string) bool,
+	scCaps *v1.Capabilities,
+	forbidAll bool,
+	exceptionCapMatchers map[string]func(string) bool,
+) *diagnostic.Diagnostic {
+	if forbidAll {
+		// User has forbidden all capabilities
+		for _, scCap := range scCaps.Add {
+			var excluded bool
+			for _, exceptionCapMatcher := range exceptionCapMatchers {
+				if exceptionCapMatcher(string(scCap)) {
+					// Forgive this capability
+					excluded = true
+					break
+				}
+			}
+			if !excluded {
+				return &diagnostic.Diagnostic{
+					Message: fmt.Sprintf(
+						"container %q has ADD capability: %q, but no capabilities should be added at all and"+
+							" this capabilty is not included in the exceptions list",
+						containerName,
+						scCap),
+				}
+			}
+		}
+		// No violations
+		return nil
+	}
+
+	// Any capability from scCaps should not match with any from paramCaps
+	for _, paramCapMatcher := range paramCapMatchers {
+		for _, scCap := range scCaps.Add {
```

Comment:
So, the tricky thing here is that containers can have some capabilities implicitly, through defaults. This means we should either:
1. Use the default list of defaults from [here](https://github.com/moby/moby/blob/master/oci/caps/defaults.go#L6-L19), and assume they are in the container.
1. Be extra-restrictive, and say that to forbid a capability, you _must_ drop it explicitly. (This is the same as assuming that containers have all capabilities by default.)

As written, the `NET_RAW` check will not fire by default on an empty capabilities list. But it should.

Reply:
Yes it will not fire on checking ADD list, but it still fires on checking DROP list right?